Source/WebCore/loader/FrameLoader.cpp

```diff
@@ -1275,8 +1275,11 @@ void FrameLoader::loadInSameDocument(URL url, RefPtr<SerializedScriptValue> stat
 
     if (document->settings().navigationAPIEnabled() && !url.isAboutBlank()) {
         if (RefPtr domWindow = document->domWindow()) {
-            if (RefPtr currentItem = checkedHistory()->currentItem())
+            if (RefPtr currentItem = checkedHistory()->currentItem()) {
+                domWindow->protectedNavigation()->dispatchNavigationEvent(url, oldURL, policyChecker().loadType());
+                // FIXME: Properly handle result of navigate event.
                 domWindow->protectedNavigation()->updateForNavigation(currentItem.releaseNonNull(), policyChecker().loadType());
```

Comment:
It looks like these two operations are quite far apart in the specification:
https://html.spec.whatwg.org/multipage/browsing-the-web.html#apply-the-history-step

Did you consider a different place for sending the navigate Event? If we can't do it right now but want to later then maybe we can add a FIXME.